src/_guides/environment-declarations.md

```diff
@@ -0,0 +1,188 @@
+---
+title: Configuring apps with compilation environment declarations
+description: >
+  Learn about using compilation environment declarations to customize application behavior.
+---
+
+Dart supports specifying compilation environment declarations when
+building or running an application.
+Compilation environment declarations are used to specify
+configuration values or options as key-value pairs
+that are accessed and evaluated at compile time.
+
+{{site.alert.note}}
+  This page uses "environment" and "environment declarations"
+  to refer to the compilation environment, not the OS environment
+  often referred to as "the environment" in software terminology.
+{{site.alert.end}}
+
+Your app can use the values of environment declarations
+to change its functionality or behavior.
+Dart compilers can take advantage of this to eliminate code
+no longer reachable due to the values specified, through "tree shaking".
+
+Some potential use cases for defining environment declarations are:
+
+* Adding functionality during debugging, such as enabling logging.
+* Creating separate flavors of your application.
+* Configuring application behavior, such as the port of an HTTP server.
+* Enabling an experimental mode of your application for testing.
+* Switching between testing and production backends.
+
+To specify an environment declaration
+to an application with the [`dart run`][] or [`dart compile`][] subcommands,
+use the `-D` or `--define` options and specify `<NAME>=<VALUE>`:
+
+```terminal
+$ dart run --define=DEBUG=true -DFLAVOR=free
+```
+
+To learn more about specifying environment declarations,
+including for other tools and compilers, see
+[Specifying environment declarations][].
+
+[`dart run`]: /tools/dart-run
+[`dart compile`]: /tools/dart-compile
+[Specifying environment declarations]: #specifying-environment-declarations
+
+## Accessing environment declarations
+
+To access specified environment declaration values,
+use one of the `fromEnvironment` constructors
+with `const` or within a constant context.
+Use [`bool.fromEnvironment`][bool-from] for `true` or `false` values,
+[`int.fromEnvironment`][int-from] for integer values,
+and [`String.fromEnvironment`][string-from] for anything else.
+
+{{site.alert.note}}
+  The environment declaration constructors are only guaranteed
+  to work when invoked as `const`.
+  Most compilers must be able to evaluate their value at compile time.
+{{site.alert.end}}
```

Comment:
**question:** Is this not redundant with the previous paragraph?

Reply:
The previous paragraph does go into similar details, but technically it's possible to try to call them not in a `const` context (even if you shouldn't). This has caused a lot of confusion and problems in the past, so I think it's worth it to be extra explicit here.